src/tribler/core/components/ipv8/eva/protocol.py

```diff
@@ -185,46 +284,74 @@ async def on_write_request_packet(self, peer: Peer, payload: WriteRequest):
+    async def on_read_request(self, peer: Peer, payload: ReadRequest):
+        logger.debug(f'On read request. Peer: {peer}. Info: {payload.info}.')
+        if peer in self.outgoing:
             return
 
-        if self._is_simultaneously_served_transfers_limit_exceeded():
-            exception = TransferLimitException('Maximum simultaneous transfers limit exceeded')
-            self._finish_with_error(transfer, exception)
+        data = b''
+        exception = None
+        if self.on_request:
+            try:
+                data = await self.on_request(peer, payload.info)
+            except Exception as e:  # pylint: disable=broad-except
+                exception = e
+        else:
+            exception = NotImplementedError('on_request callback is not defined')
+
+        transfer = OutgoingTransfer(
+            container=self.outgoing,
+            peer=peer,
+            info=payload.info,
+            data=data,
+            data_size=len(data),
+            nonce=payload.nonce,
+            settings=self.settings,
+            parent_task_group=self.task_group,
+            send_message=self.send_message,
+            on_complete=self.on_send_complete,
+            on_error=self.on_error
+        )
+
+        if exception:
+            message = f'{exception.__class__.__name__}: {exception}'
+            logger.warning(f'Exception during data retrieval. {message}')
+            self._finish_with_error(transfer, RequestRejected(message, transfer))
             return
 
-        self.incoming[peer] = transfer
-
-        self.community.register_anonymous_task('eva_terminate_by_timeout', transfer.terminate_by_timeout_task)
-        self.community.register_anonymous_task('eva_resend_acknowledge', self._resend_acknowledge_task, transfer)
-        self.send_message(peer, transfer.make_acknowledgement())
+        if self.check_transfer_correctness(transfer):
+            transfer.start()
 
     @message_handler(Acknowledgement)
     async def on_acknowledgement_packet(self, peer: Peer, payload: Acknowledgement):
         logger.debug(f'On acknowledgement({payload.number}). Window size: {payload.window_size}. Peer: {peer}.')
 
-        transfer = self.outgoing.get(peer)
+        transfer = self._get_transfer(peer=peer, container=self.outgoing, nonce=payload.nonce)
         if not transfer:
-            logger.warning(f'No outgoing transfer found with peer {peer} associated with incoming acknowledgement.')
-            return
-
-        if transfer.nonce != payload.nonce:
-            logger.warning(f'Cannot handle incoming acknowledgement from peer {peer} - nonce mismatch.')
             return
 
         data_list = list(transfer.on_acknowledgement(payload.number, payload.window_size))
         is_transfer_finished = not data_list
         if is_transfer_finished:
-            self.send_scheduled()
+            transfer.finish(result=transfer.create_result())
+            # self.scheduler.send_scheduled()
```

Comment:
I think it's a time to remove this and other similar commented out blocks of code :)

Reply:
Fixed, thank you!